Use Boto3 for ec2_group Fixes #23507

sramakr opened the pull request:
##### SUMMARY

 Currently boto doesn't support ipv6. To support ipv6 in ec2_group, we need boto3.
boto3 has significant API changes, which caused more re-factoring for ec2_group module.
Added additional integration test to test_ec2_group role. 
Fixes - ec2_group IPv6 rule fails with error CIDR is malformed #23507


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
ec2_group
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0
  config file =
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /vagrant/myenv/local/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /vagrant/myenv/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```


##### ADDITIONAL INFORMATION
Upgrade ec2_group module to support boto3.
boto3 security group has significant change in it's API's for SecurityGroup. For starters, it always handles in Json. Rules are handled through IpPermissions.This PR, constructs IpPermissions for both ipv4 and ipv6.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
PARSED OUTPUT
{
    "changed": true,
    "group_id": "sg-a82a40d3",
    "invocation": {
        "module_args": {
            "aws_access_key": "XXXXX",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "description": "hello",
            "ec2_access_key": "XXXX",
            "ec2_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ec2_url": null,
            "name": "firstone",
            "profile": null,
            "purge_rules": true,
            "purge_rules_egress": true,
            "region": "us-west-2",
            "rules": [
                {
                    "cidr_ip": [
                        "172.16.1.0/24",
                        "172.16.17.0/24"
                    ],
                    "cidr_ipv6": [
                        "X:X:X::/40"
                    ],
                    "ports": 5665,
                    "proto": "tcp"
                }
            ],
            "rules_egress": [
                {
                    "cidr_ipv6": "X:X:X::/40",
                    "ports": 8085,
                    "proto": "tcp"
                }
            ],
            "security_token": null,
            "state": "present",
            "validate_certs": true,
            "vpc_id": "vpc-54d1c236"
        }
    }
}
```

sramakr committed:
commit with better code

sramakr committed:
Use Boto3 for ec2_group

Currently boto doesn't support ipv6. To support ipv6 in ec2_group, we need boto3.
boto3 has significant API changes, which caused more re-factoring for ec2_group module.
Added additional integration test to test_ec2_group role.

sramakr commented:
@s-hertel PR for using boto3 in ec2_group. appreciate your feedback. thanks.

willthames commented on lib/ansible/module_utils/ec2.py:
```diff
@@ -94,9 +94,10 @@ def found(response_code):
-    except ValueError:
-        module.fail_json(msg='There is an issue in the code of the module. You must specify either both, resource or client to the conn_type '
-                             'parameter in the boto3_conn function call')
+    except AnsibleAWSError, awsError:
```
when do we expect awsError?

edit: oops, got confused by old school python exception syntax. Use new exception syntax (`as e`)

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -174,12 +174,13 @@
 import time
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import ec2_connect, ec2_argument_spec
+from ansible.module_utils import ec2 as ec2_utils
```
Don't do this, import the specific things you need.

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -383,37 +381,57 @@ def main():
     state = module.params.get('state')
     purge_rules = module.params['purge_rules']
     purge_rules_egress = module.params['purge_rules_egress']
+    profile_name="myprofile"
```
Please be careful what you commit.

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -383,37 +381,57 @@ def main():
     changed = False
 
-    ec2 = ec2_connect(module)
+    #ec2 = ec2_connect(module)
```
just delete things rather than comment them out

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -174,12 +174,13 @@
-    from boto.ec2.securitygroup import SecurityGroup
-    from boto.exception import BotoServerError
+    import boto3
+    from botocore import exceptions
```
`import botocore` will make this code a lot more consistent with most other boto3-based ansible modules

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -479,44 +496,57 @@ def main():
+                                try:
+                                    group.authorize_ingress(GroupName=group.group_name, IpPermissions=[ips])
+                                except exceptions.ClientError as e:
+                                    module.fail_json(msg="Unable to authorize ingress for group %s security group '%s' - %s" % (group_id, group, e.message), exception=traceback.format_exc())
```
Use the Ansible standard [exception handling guidelines for boto3](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-2)

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -531,45 +561,52 @@ def main():
+                            if ipPermissions:
+                                ips=ipPermissions
+                                if vpc_id:
+                                    ips=map(lambda x: map(lambda y: y.update({'VpcId': vpc_id}), x.get('UserIdGroupPairs')), [ipPermissions])
```
This code is pretty incomprehensible, I'd try and avoid map and lambda and use a list/dictionary comprenhension (or a for loop if it's still unreadable)

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -597,5 +626,61 @@ def main():
         module.exit_json(changed=changed, group_id=None)
 
 
+def createGrantGroupIpPermission(group_id, rule):
```
don't add methods after `main()`, it's very unexpected!

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -578,17 +615,9 @@ def main():
-                        to_port=rule.to_port,
-                        src_group_id=grantGroup,
-                        cidr_ip=grant.cidr_ip)
+                    group.revoke_egress(GroupId=group.id, IpPermissions=[ipPermissions])
```
Using http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Client.revoke_security_group_egress is more usual for most of our boto3 module (and more similar to the existing code). This is just one example, the same applies for other actions on the `SecurityGroup` class

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -383,37 +381,57 @@ def main():
+        response = client.describe_security_groups()
+        if 'SecurityGroups' in response:
+            security_groups = response.get('SecurityGroups')
+    except exceptions.NoCredentialsError as e:
```
Once we have a client, I don't think we need to handle this exception again.

sramakr replied:
Client is returned as not none when a valid region is specified and aws key and secret are None. So, describe security groups fails with NoCredentialsError and also the existing integration test.

willthames reviewed (changes requested):
This PR would benefit from some reading of existing boto3 modules, there is a reasonable consistency, particularly of late, and there are a lot of changes here that are surprising if you're used to the existing codebase. So apologies if I come across as unnecessarily picky.

sramakr commented:
@willthames Thanks for the comments. I have to admit this is my first PR in ansible, sorry , it's out of standard guidelines. I will fix and update the PR.

sramakr committed:
Follow the standard for boto3 ansible

Fixed imports. Use boto3 ansible exception with camel_dict_to_snake_dict.
Refactored the call to authorize/revoke  ingress and egress.

s-hertel commented:
Hi @sramakr, thanks for the work you've put into this.

I haven't tested it out yet (it's on my list), but I'm a little nervous about changing the exception handling in lib/ansible/module_utils/ec2.py since so many things use it. Is there a bug you're fixing? For the changes to how security_token is being set, I think you can discard it since there was a change merged recently that should fix when security_token is an empty string (which appears to be what you're fixing there).

Also, could you rebase on devel please? :) Looks like there are conflicts.

sramakr commented:
@s-hertel Found couple of issues

There was no validation to check if the region is a valid ec2 region or not.
```
if region not in boto3.session.Session(profile_name=profile).get_available_regions(resource):
        raise AnsibleAWSError("Region %s does not seem to be available for aws module %s. If the region definitely exists, you may need to upgrade "
                              "boto or extend with endpoints_path" % (region, 'boto.ec2'))
```

Value error was hard-coded in the exception handling , below change will return the actual error message from ec2 , instead of the hard-coded message of 'There is an issue in the calling code. You ...'

```
try:
        return _boto3_conn(conn_type=conn_type, resource=resource, region=region, endpoint=endpoint, **params)
    except AnsibleAWSError as awsError:
        module.fail_json(msg=awsError.message)
    except ValueError as valueError:
        module.fail_json(msg=valueError.message)
```
hmm... I am not sure about the security token , I will remove it and rebase and update the PR>

sramakr committed:
Reverted back changes to ec2.py

sramakr committed:
Formatted the code

sramakr committed:
Removed dependancy with module ipaddress

Added new parameter called cidr_ipv6 for specifying
ipv6 addresses inline with how boto3 handles ipv6 addresses.

sramakr committed:
Rebased with devel

Fixed formatting error

sramakr committed:
Updated integration test

sramakr committed:
Removing print statements

sramakr committed:
Added ipv6 integration test for ec2_group

sramakr committed:
Set purge_rules to false for integration test

sramakr committed:
Fixed more tests for ipv6

sramakr commented:
@s-hertel  @willthames I added more tests and cleaned up my code. But, some of my tests are failing with unable to authorize errors , especially when I add rules for ipv6 or ipv4. Should I remove those tests? 

Appreciate any help.
Thanks
Sriram

s-hertel commented:
@sramakr They're all passing locally for me and look good. We'll need to add some permissions for the CI to pass, but that's fine. No need to remove them!

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -178,15 +177,18 @@
+    import boto3
+    import botocore
+
+    HAS_BOTO3 = True
```
You can import HAS_BOTO3 from ansible.module_utils.ec2 so you don't need to set the value here or in the except ImportError.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -19,7 +19,6 @@
                     'status': ['stableinterface'],
                     'supported_by': 'curated'}
 
-
```
Can you add requirements to the docs above and add boto3 to that? (here's an example: https://github.com/ansible/ansible/blob/2be64b8c3af2caf8fc72283a02dd33cc3f636c6c/lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway_facts.py#L28)

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -178,15 +177,18 @@
 import time
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import ec2_connect, ec2_argument_spec
+import ansible.module_utils.ec2 as ec2_utils
```
We shouldn't need this import. You'll need to explicitly import boto3_conn (and possible others) but that's fairly normal

willthames replied:
This is still unusual code, and my review stands.

willthames reviewed (commented):
Thanks for all your hard work on this, I really appreciate the effort you've made to get this to meet standards!

willthames commented:
I'll be testing this when I get a chance in the next couple of days, just in case I'm doing anything missed by tests.

s-hertel reviewed (approved):
I've been testing this and it looks good to me. The integration tests are all passing locally as well.

sramakr committed:
Fixed import statements

Added example for ipv6.
Removed defining HAS_BOTO3 variable and import HAS_BOTO3 from ec2.
Cleaned up import statements.

gundalow commented:
CI is failing with
```
2017-06-13 01:47:18 TASK [ec2_group : test state=present for ipv6 (expected changed=true)] *********
2017-06-13 01:47:18 task path: /root/ansible/test/integration/targets/ec2_group/tasks/main.yml:232
2017-06-13 01:47:18 Using module file /root/ansible/lib/ansible/modules/cloud/amazon/ec2_group.py
2017-06-13 01:47:18 <127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
2017-06-13 01:47:18 <127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
2017-06-13 01:47:18 <127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459 `" && echo ansible-tmp-1497318438.15-116449959747459="` echo /root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459 `" ) && sleep 0'
2017-06-13 01:47:18 <127.0.0.1> PUT /tmp/tmpbseLt9 TO /root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459/ec2_group.py
2017-06-13 01:47:18 <127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459/ /root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459/ec2_group.py && sleep 0'
2017-06-13 01:47:18 <127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-aVMepA/coverage/injector.py /root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459/ec2_group.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1497318438.15-116449959747459/" > /dev/null 2>&1 && sleep 0'
2017-06-13 01:47:18 The full traceback is:
2017-06-13 01:47:18 Traceback (most recent call last):
2017-06-13 01:47:18   File "/tmp/ansible_M2Qiwg/ansible_module_ec2_group.py", line 404, in authorize_ip
2017-06-13 01:47:18     IpPermissions=[ip_permission])
2017-06-13 01:47:18   File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 253, in _api_call
2017-06-13 01:47:18     return self._make_api_call(operation_name, kwargs)
2017-06-13 01:47:18   File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 557, in _make_api_call
2017-06-13 01:47:18     raise error_class(parsed_response, operation_name)
2017-06-13 01:47:18 ClientError: An error occurred (UnauthorizedOperation) when calling the AuthorizeSecurityGroupIngress operation: You are not authorized to perform this operation. Encoded authorization failure message: 17ZaTR3W4hMkMjoyJhnf08GwS_VENY7kWj5pwiJ52xxPXUPwPEa9VFqMuZhm5jHJdRV14BajPhuULd5BPBQ7aWjXmTCcPkoe8mfO433kJ_TrMuTSmg4O_WeTPgHT8ZfUeyT7ptlB6JG6M6VyZFUYgn4WvYa6MA8lVIJCCFLBhoEdsX95F-N6hCcvU8OeY2U02FjUmPbuRi8UOHzw11V_PLhqz5_L86L_2YoFBYBcCaonSUHsLuDp4Kb95MxqZZX82j9GgSD07-voiPvwDlfb1_tQFVmdKmCmUYAKx9Jh22vNMlmAB9KB6Y8sMu1DTehukNaqYcGh7E_J2BtUxJBLGKYrXiyQMxXZoqCwoY4Hd0BaSrrq8vW0DJ7BSBQgkjr6tsXDmZ2uACY3cDxJj67lUBeXcR2Si-myOUDSVYJtwIQApA-7CorEEU_tMqM2h5qy106dLGxU_PmN0lTxj2-MkJpc3Cu3L5Co3ID515cOfubV6zLBv6AKTwGpTA-DsRJvtMVl9UkmPSLwybFaCC5dv6DGx93IuRoIFA
2017-06-13 01:47:18 
2017-06-13 01:47:18 fatal: [localhost]: FAILED! => {
2017-06-13 01:47:18     "changed": false, 
2017-06-13 01:47:18     "error": {
2017-06-13 01:47:18         "code": "UnauthorizedOperation", 
2017-06-13 01:47:18         "message": "You are not authorized to perform this operation. Encoded authorization failure message: 17ZaTR3W4hMkMjoyJhnf08GwS_VENY7kWj5pwiJ52xxPXUPwPEa9VFqMuZhm5jHJdRV14BajPhuULd5BPBQ7aWjXmTCcPkoe8mfO433kJ_TrMuTSmg4O_WeTPgHT8ZfUeyT7ptlB6JG6M6VyZFUYgn4WvYa6MA8lVIJCCFLBhoEdsX95F-N6hCcvU8OeY2U02FjUmPbuRi8UOHzw11V_PLhqz5_L86L_2YoFBYBcCaonSUHsLuDp4Kb95MxqZZX82j9GgSD07-voiPvwDlfb1_tQFVmdKmCmUYAKx9Jh22vNMlmAB9KB6Y8sMu1DTehukNaqYcGh7E_J2BtUxJBLGKYrXiyQMxXZoqCwoY4Hd0BaSrrq8vW0DJ7BSBQgkjr6tsXDmZ2uACY3cDxJj67lUBeXcR2Si-myOUDSVYJtwIQApA-7CorEEU_tMqM2h5qy106dLGxU_PmN0lTxj2-MkJpc3Cu3L5Co3ID515cOfubV6zLBv6AKTwGpTA-DsRJvtMVl9UkmPSLwybFaCC5dv6DGx93IuRoIFA"
2017-06-13 01:47:18     }, 
2017-06-13 01:47:18     "failed": true, 
2017-06-13 01:47:18     "invocation": {
2017-06-13 01:47:18         "module_args": {
2017-06-13 01:47:18             "aws_access_key": "ASIAICKO5SCOSEULDQBA", 
2017-06-13 01:47:18             "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
2017-06-13 01:47:18             "description": "Created by ansible integration tests", 
2017-06-13 01:47:18             "ec2_access_key": "ASIAICKO5SCOSEULDQBA", 
2017-06-13 01:47:18             "ec2_region": "us-east-1", 
2017-06-13 01:47:18             "ec2_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
2017-06-13 01:47:18             "ec2_url": null, 
2017-06-13 01:47:18             "name": "shippable-25271-40", 
2017-06-13 01:47:18             "profile": null, 
2017-06-13 01:47:18             "purge_rules": true, 
2017-06-13 01:47:18             "purge_rules_egress": true, 
2017-06-13 01:47:18             "region": "us-east-1", 
2017-06-13 01:47:18             "rules": [
2017-06-13 01:47:18                 {
2017-06-13 01:47:18                     "cidr_ipv6": "64:ff9b::/96", 
2017-06-13 01:47:18                     "from_port": 8182, 
2017-06-13 01:47:18                     "proto": "tcp", 
2017-06-13 01:47:18                     "to_port": 8182
2017-06-13 01:47:18                 }
2017-06-13 01:47:18             ], 
2017-06-13 01:47:18             "rules_egress": null, 
2017-06-13 01:47:18             "security_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
2017-06-13 01:47:18             "state": "present", 
2017-06-13 01:47:18             "validate_certs": true, 
2017-06-13 01:47:18             "vpc_id": null
2017-06-13 01:47:18         }
2017-06-13 01:47:18     }, 
2017-06-13 01:47:18     "msg": "Unable to authorize in for ip 64:ff9b::/96 security group 'shippable-25271-40' - An error occurred (UnauthorizedOperation) when calling the AuthorizeSecurityGroupIngress operation: You are not authorized to perform this operation. Encoded authorization failure message: 17ZaTR3W4hMkMjoyJhnf08GwS_VENY7kWj5pwiJ52xxPXUPwPEa9VFqMuZhm5jHJdRV14BajPhuULd5BPBQ7aWjXmTCcPkoe8mfO433kJ_TrMuTSmg4O_WeTPgHT8ZfUeyT7ptlB6JG6M6VyZFUYgn4WvYa6MA8lVIJCCFLBhoEdsX95F-N6hCcvU8OeY2U02FjUmPbuRi8UOHzw11V_PLhqz5_L86L_2YoFBYBcCaonSUHsLuDp4Kb95MxqZZX82j9GgSD07-voiPvwDlfb1_tQFVmdKmCmUYAKx9Jh22vNMlmAB9KB6Y8sMu1DTehukNaqYcGh7E_J2BtUxJBLGKYrXiyQMxXZoqCwoY4Hd0BaSrrq8vW0DJ7BSBQgkjr6tsXDmZ2uACY3cDxJj67lUBeXcR2Si-myOUDSVYJtwIQApA-7CorEEU_tMqM2h5qy106dLGxU_PmN0lTxj2-MkJpc3Cu3L5Co3ID515cOfubV6zLBv6AKTwGpTA-DsRJvtMVl9UkmPSLwybFaCC5dv6DGx93IuRoIFA", 
2017-06-13 01:47:18     "response_metadata": {
2017-06-13 01:47:18         "http_headers": {
2017-06-13 01:47:18             "date": "Tue, 13 Jun 2017 01:47:17 GMT", 
2017-06-13 01:47:18             "server": "AmazonEC2", 
2017-06-13 01:47:18             "transfer-encoding": "chunked"
2017-06-13 01:47:18         }, 
2017-06-13 01:47:18         "http_status_code": 403, 
2017-06-13 01:47:18         "request_id": "6d0f9432-574e-413e-a7d0-ec4ddd8f0d8c", 
2017-06-13 01:47:18         "retry_attempts": 0
2017-06-13 01:47:18     }
2017-06-13 01:47:18 }
```

s-hertel commented:
@willthames Will you rereview/clear you requested changes on this, please?

mattclay commented:
@sramakr Please don't remove the tests. I'll be updating the IAM permission to resolve the issue.

sramakr commented:
@mattclay ok. I have reverted the last commit. Thanks.

mattclay commented:
The current CI failure appears to be an issue with exception handling:

```
2017-06-22 17:38:39 The full traceback is:
2017-06-22 17:38:39 Traceback (most recent call last):
2017-06-22 17:38:39   File "/tmp/ansible_r3Z_XF/ansible_module_ec2_group.py", line 739, in <module>
2017-06-22 17:38:39     main()
2017-06-22 17:38:39   File "/tmp/ansible_r3Z_XF/ansible_module_ec2_group.py", line 512, in main
2017-06-22 17:38:39     module.fail_json(msg="Error in describe_security_groups: %s" % e.msg, exception=traceback.format_exc(),
2017-06-22 17:38:39 AttributeError: 'ClientError' object has no attribute 'msg'
2017-06-22 17:38:39 
2017-06-22 17:38:39 fatal: [localhost]: FAILED! => {
2017-06-22 17:38:39     "changed": false, 
2017-06-22 17:38:39     "failed": true, 
2017-06-22 17:38:39     "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_r3Z_XF/ansible_module_ec2_group.py\", line 739, in <module>\n    main()\n  File \"/tmp/ansible_r3Z_XF/ansible_module_ec2_group.py\", line 512, in main\n    module.fail_json(msg=\"Error in describe_security_groups: %s\" % e.msg, exception=traceback.format_exc(),\nAttributeError: 'ClientError' object has no attribute 'msg'\n", 
2017-06-22 17:38:39     "module_stdout": "", 
2017-06-22 17:38:39     "msg": "MODULE FAILURE", 
2017-06-22 17:38:39     "rc": 0
2017-06-22 17:38:39 }
```

Hopefully once this is fixed it will expose the actual error.

mattclay commented:
@sramakr It looks like you need to update the assertion that's failing:

```
"assertion": "\"validate the provided access credentials\" in result.msg",
```

It should probably be:

```
"assertion": "\"validate the provided access credentials\" in result.error.message",
```

However, it might better to keep the exception message in the error. You'll just have to check to make sure there's actually a message in the exception that you can use. I'd check with @s-hertel to see what the recommended approach is.

willthames commented:
Looking at all the other boto3 exception handling, we clearly have no common approach to this, although I know @mikedlr is working on some common code that can be used for all boto3 exception handling.

I think I'm fine with your current approach, you just need to fix the test.

sramakr committed:
Fixed exception handling

sramakr commented:
Updated exception to handle both python 3 and 2.7. Not sure if there is a better way , but had to use the exception object (e ) in the message , instead of e.message, now it works for both Python 2.7 and 3. I agree , need a common approach to handle the exceptions.
I guess now I just need permissions to run these tests. once we have that, I think we are good to go.

Commit:
Add IAM permissions for ec2_group tests

Missing AuthorizeSecurityGroupEgress necessary for latest tests

willthames commented:
I have rebased this and reran the tests locally, after also updating the EC2 IAM policy to allow SecurityGroupEgress.

Once tests pass I'll merge this.

Commit:
Wrapped botocore import in try/except block

Import just botocore to be more similar to other modules

willthames commented:
All expected tests passing, so merged this. @mattclay, IAM permissions need updating to match https://github.com/ansible/ansible/blob/devel/hacking/aws_config/testing_policies/ec2-policy.json#L37-L38

willthames commented:
Thanks for this contribution @sramakr, it'll really help us move ec2_group module forward.

mattclay commented:
@willthames Please do not merge PRs which fail to pass tests. Doing so breaks all future merge CI runs, and often many (sometimes all) PR CI runs as well.

willthames commented:
@mattclay ack, sorry.

mattclay commented:
The policy has been updated.
